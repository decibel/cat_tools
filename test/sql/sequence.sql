\set ECHO none

\i test/setup.sql

-- test_role is set in test/deps.sql

SET LOCAL ROLE :use_role;

CREATE TEMP TABLE seqtest(s serial);
CREATE TEMP TABLE functions(fname name);
INSERT INTO functions SELECT unnest(
  '{currval, setval, nextval, sequence__next, sequence__last, sequence__set_next, sequence__set_last}'::text[]
);
CREATE TEMP VIEW fv AS
  SELECT 'cat_tools'::name AS sname, fname
      , (SELECT proargtypes::regtype[] FROM pg_proc WHERE oid=('cat_tools.' || fname)::regproc) AS args
    FROM functions
;

SELECT plan((
  0
  + (SELECT count(*) FROM functions) -- no_use tests
  + 4 * (SELECT count(*) FROM functions) -- definition

  +2 -- unset
  +2 -- next
  +8 -- set
  +3 -- currval
)::int);

GRANT SELECT ON functions TO public;
SET LOCAL ROLE :no_use_role;
SELECT throws_ok(
  format( $$SELECT cat_tools.%I('seqtest', 's')$$, fname )
  , '42501' -- insufficient_privilege
  , NULL
  , format( 'Permission denied trying to use cat_tools.%I()', fname )
)
  FROM functions
;

SET LOCAL ROLE :use_role;

SELECT function_returns( sname, fname, 'bigint'::regtype::text ) FROM fv;
SELECT volatility_is( sname, fname, 'volatile' ) FROM fv;
SELECT isnt_strict( sname, fname ) FROM fv;
SELECT isnt_definer( sname, fname ) FROM fv;

/*
 * unset
 */
SELECT throws_ok(
  format( $$SELECT cat_tools.%I('seqtest', 's')$$, fname )
  , '55000' -- object_not_in_prerequisite_state
  , NULL
  , 'Call on unset sequence throws error'
)
  FROM functions
  WHERE fname ~ 'currval|__last'
;

/*
 * next
 */
SELECT is(
  cat_tools.nextval('seqtest', 's')
  , 1::bigint
  , 'nextval returns correct value'
);
SELECT is(
  cat_tools.sequence__next('seqtest', 's')
  , 2::bigint -- Previous call to nextval incremented this...
  , 'sequence__next returns correct value'
);

/*
 * set
 */
SELECT is(
  cat_tools.setval('seqtest', 's', 2)
  , 2::bigint
  , 'setval returns correct value'
);
SELECT is(
  cat_tools.sequence__next('seqtest', 's')
  , 3::bigint
  , 'sequence__next returns correct value'
);
SELECT is(
  cat_tools.setval('seqtest', 's', 4, false)
  , 4::bigint
  , 'setval returns correct value'
);
SELECT is(
  cat_tools.sequence__next('seqtest', 's')
  , 4::bigint
  , 'sequence__next returns correct value'
);
SELECT is(
  cat_tools.sequence__set_last('seqtest', 's', 5)
  , 5::bigint
  , 'sequence__set_last returns correct value'
);
SELECT is(
  cat_tools.sequence__next('seqtest', 's')
  , 6::bigint
  , 'sequence__next returns correct value'
);
SELECT is(
  cat_tools.sequence__set_next('seqtest', 's', 7)
  , 7::bigint
  , 'sequence__set_next returns correct value'
);
SELECT is(
  cat_tools.sequence__next('seqtest', 's')
  , 7::bigint
  , 'sequence__next returns correct value'
);

/*
 * currval
 */
SELECT lives_ok(
  'INSERT INTO seqtest VALUES(default)'
  , 'INSERT INTO seqtest VALUES(default)'
);
SELECT is(
  cat_tools.currval('seqtest', 's')
  , 8::bigint
  , 'currval'
);
SELECT is(
  cat_tools.sequence__last('seqtest', 's')
  , 8::bigint
  , 'currval'
);

\i test/pgxntool/finish.sql

-- vi: expandtab ts=2 sw=2
