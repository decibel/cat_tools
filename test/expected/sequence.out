\set ECHO none
1..50
ok 1 - Permission denied trying to use cat_tools.currval()
ok 2 - Permission denied trying to use cat_tools.setval()
ok 3 - Permission denied trying to use cat_tools.nextval()
ok 4 - Permission denied trying to use cat_tools.sequence__next()
ok 5 - Permission denied trying to use cat_tools.sequence__last()
ok 6 - Permission denied trying to use cat_tools.sequence__set_next()
ok 7 - Permission denied trying to use cat_tools.sequence__set_last()
ok 8 - Function cat_tools.currval() should return bigint
ok 9 - Function cat_tools.setval() should return bigint
ok 10 - Function cat_tools.nextval() should return bigint
ok 11 - Function cat_tools.sequence__next() should return bigint
ok 12 - Function cat_tools.sequence__last() should return bigint
ok 13 - Function cat_tools.sequence__set_next() should return bigint
ok 14 - Function cat_tools.sequence__set_last() should return bigint
ok 15 - Function cat_tools.currval() should be VOLATILE
ok 16 - Function cat_tools.setval() should be VOLATILE
ok 17 - Function cat_tools.nextval() should be VOLATILE
ok 18 - Function cat_tools.sequence__next() should be VOLATILE
ok 19 - Function cat_tools.sequence__last() should be VOLATILE
ok 20 - Function cat_tools.sequence__set_next() should be VOLATILE
ok 21 - Function cat_tools.sequence__set_last() should be VOLATILE
ok 22 - Function cat_tools.currval() should not be strict
ok 23 - Function cat_tools.setval() should not be strict
ok 24 - Function cat_tools.nextval() should not be strict
ok 25 - Function cat_tools.sequence__next() should not be strict
ok 26 - Function cat_tools.sequence__last() should not be strict
ok 27 - Function cat_tools.sequence__set_next() should not be strict
ok 28 - Function cat_tools.sequence__set_last() should not be strict
ok 29 - Function cat_tools.currval() should not be security definer
ok 30 - Function cat_tools.setval() should not be security definer
ok 31 - Function cat_tools.nextval() should not be security definer
ok 32 - Function cat_tools.sequence__next() should not be security definer
ok 33 - Function cat_tools.sequence__last() should not be security definer
ok 34 - Function cat_tools.sequence__set_next() should not be security definer
ok 35 - Function cat_tools.sequence__set_last() should not be security definer
ok 36 - Call on unset sequence throws error
ok 37 - Call on unset sequence throws error
ok 38 - nextval returns correct value
ok 39 - sequence__next returns correct value
ok 40 - setval returns correct value
ok 41 - sequence__next returns correct value
ok 42 - setval returns correct value
ok 43 - sequence__next returns correct value
ok 44 - sequence__set_last returns correct value
ok 45 - sequence__next returns correct value
ok 46 - sequence__set_next returns correct value
ok 47 - sequence__next returns correct value
ok 48 - INSERT INTO seqtest VALUES(default)
ok 49 - currval
ok 50 - currval
# TRANSACTION INTENTIONALLY LEFT OPEN!
